### ISSUE TITLE
ActionType doesn't exclude non-action creators anymore

### ISSUE THREAD

**Issue by voliva** [piotrwitek/typesafe-actions#132]
## Description
I was using ActionType to build the union of all of my action creators by doing:

```ts
// user.ts
export enum UserActionTypes {
    SomeAction = 'UserSomeAction'
}

export const someAction = createStandardAction(UserActionTypes.Someaction)<string>();
```

```ts
// index.ts
import * as userActions from './user';

type UserActions = ActionType<typeof userActions>;

// UserActions = PayloadAction<.....>
```

This used to work on TS@3.3 and typesafe-actions@3.0.0, but it doesn't anymore in the last version. Now UserActions is `userActions.UserActionTypes | PayloadAction<....>`

## Suggested solution(s)

I think that by adding an additional check into the ActionType this would get solved:

```ts
type ActionType<ActionCreatorOrMap> = ActionCreatorOrMap extends ActionCreator<string>
    ? ReturnType<ActionCreatorOrMap>
    : ActionCreatorOrMap extends object
    ? {
          [K in keyof ActionCreatorOrMap]: ActionCreatorOrMap[K] extends ActionCreator<any>
              ? ActionType<ActionCreatorOrMap[K]>
              : never
      }[keyof ActionCreatorOrMap]
    : ActionCreatorOrMap;
```

This is similar to pre-3.3.0, but based on it.... I see that change was added to solve a circular dependency after upgrading to TS@3.4.1 - I haven't had any errors/warnings by using this version of ActionType.

Should I create a PR and move the discussion there, if this change makes sense?

## Project Dependencies
- Typesafe-Actions Version: 3.3.0
- TypeScript Version: 3.4.1

**Comment by piotrwitek** [comment]
The suggested solution is not acceptable because it only works for nesting 1 level deep and adds unnecessary computational complexity, the original is a recursive type working for deeply nested structures.

The problem is within the compiler and `never` cannot be used to remove unmatched types in such conditional type anymore. Hopefully, it'll be fixed in the future hotfix for TypeScript compiler so we can bring back 3.3.0 behaviour.

**Comment by voliva** [comment]
Ok, makes sense - either I'll move those enums out or wait for a TSC fix before upgrading.

Thank you!

**Comment by piotrwitek** [comment]
@voliva I think I found temporary fix to compiler issue, please check this type on your codebase and report back to me if that fixed your issues: 
```ts
export type ActionType<
  ActionCreatorOrMap
> = ActionCreatorOrMap extends ActionCreator<string>
  ? ReturnType<ActionCreatorOrMap>
  : ActionCreatorOrMap extends object
  ? {
      [K in keyof ActionCreatorOrMap]: ActionType<ActionCreatorOrMap[K]>
    }[keyof ActionCreatorOrMap]
  : ActionCreatorOrMap extends infer R // should be just never but compiler yell with circularly references itself error
  ? never
  : never;
```

**Comment by voliva** [comment]
Hey, thanks for taking a deeper look on this - Unfortunately it doesn't.

I created a codeSandbox here so we can try: https://codesandbox.io/s/z68jnxwko3

There it uses current ActionType@3.3.0, your suggestion, and the workaround I did for my project, as we're using shallow actionCreators (or just the ones returned by createAsyncAction, which return only 1-level deep, which is to define another type:

```ts
type Function = (...args: Array<any>) => any;
type FunctionOrFunctionMap = Function | { [key: string]: Function };
type FunctionsOf<Module> = {
  [K in keyof Module]: Module[K] extends FunctionOrFunctionMap
    ? Module[K]
    : never
}[keyof Module];
```

And then using this with ActionType as

```ts
type UserActions = ActionType<FunctionsOf<typeof userActions>>;
```

This workaround does cover my current setup.

**Comment by piotrwitek** [comment]
@voliva you must be doing something wrong, I have used the code from your sandbox in my local env and the solution I posted working perfectly

**Comment by piotrwitek** [comment]
Fixed in v3.4.0 release

**Comment by voliva** [comment]
Yep, that worked perfectly - Sorry I took this long to answer.

Thank you very much!
